File: leetcode/879.盈利计划.java
/*
 * @lc app=leetcode.cn id=879 lang=java
 *
 * [879] 盈利计划
 */

// @lc code=start
class Solution {
    public int profitableSchemes(int n, int minProfit, int[] group, int[] profit) {
        // int len = group.length, MOD = (int)1e9 + 7;
        // int[][][] dp = new int[len + 1][n + 1][minProfit + 1];
        // dp[0][0][0] = 1;
        // for (int i = 1; i <= len; i++) {
        //     int members = group[i - 1], earn = profit[i - 1];
        //     for (int j = 0; j <= n; j++) {
        //         for (int k = 0; k <= minProfit; k++) {
        //             if (j < members) {
        //                 dp[i][j][k] = dp[i - 1][j][k];
        //             } else {
        //                 dp[i][j][k] = (dp[i - 1][j][k] + dp[i - 1][j - members][Math.max(0, k - earn)]) % MOD;
        //             }
        //         }
        //     }
        // }
        // int sum = 0;
        // for (int j = 0; j <= n; j++) {
        //     sum = (sum + dp[len][j][minProfit]) % MOD;
        // }
        // return sum;
        int len = group.length, Mod=(int)1e9+7;
        int[][] dp = new int[len+1][minProfit+1];
        for (int i = 0; i < len; i++) {
            int members = group[i];
            int earn = profit[i];
            for (int j = n; j >= members; j--) {
                for (int k = minProfit; k >= 0 ; k--) {
                    dp[j][k] = (dp[j][k]+dp[j-members][Math.max(0, k-earn)])% Mod;
                }
            }
        }
        return dp[n][minProfit];
    }
}
// @lc code=end

File: leetcode/golang/879.盈利计划.go
/*
 * @lc app=leetcode.cn id=879 lang=golang
 *
 * [879] 盈利计划
 */

// @lc code=start
func profitableSchemes(n int, minProfit int, group []int, profit []int) int {
	const mod int = 1e9 + 7
    dp := make([][]int, n+1)
    for i := range dp {
        dp[i] = make([]int, minProfit+1)
        dp[i][0] = 1
    }
	// for i := 0 ; i < len(group) ; i++ {
	// 	members , earn := group[i] , profit[i]
	// 	for j := n; j < count; j++ {	
	// 	}
	// }
	for i, members := range group {
        earn := profit[i]
        for j := n; j >= members; j-- {
            for k := minProfit; k >= 0; k-- {
                dp[j][k] = (dp[j][k] + dp[j-members][max(0, k-earn)]) % mod
            }
        }
    }
    return dp[n][minProfit]

}

func max (a,b int) int {
	if a > b{
		return a
	}
	return b
}
// @lc code=end

